Make sure that management router is reloaded

marko-bekhta opened the pull request:
fixes https://github.com/quarkusio/quarkus/issues/41315

I'm opening this as a draft only, as that's probably a bad idea for the fix 🙈 😕 but then ... it helps.

marko-bekhta committed:
Make sure that management router is reloaded

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -567,7 +581,8 @@ public void handle(RoutingContext event) {
 
-            VertxHttpRecorder.managementRouter = handler;
+            VertxHttpRecorder.managementRouterDelegate = handler;
+            VertxHttpRecorder.managementRouter = e -> VertxHttpRecorder.managementRouterDelegate.handle(e);
```
Probably better avoid the lambda here.

cescoffier reviewed (approved):
I can't see anything wrong. Not sure it helps, except removing routes during the hot reload.

marko-bekhta commented:
Hey Clement 😃 👋🏻,

Thanks for taking a look at this one. 

> Not sure it helps, except removing routes during the hot reload.

I've been testing it on the search.quarkus.io app and the change helped there, both with adding/removing management endpoints and with modifying the handler.

I found this test for live-reload https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/LiveReloadArtifactTest.java, so will try to add some tests using a similar approach to cover some of the management endpoint reload scenarios.

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -546,6 +547,19 @@ public void handle(RoutingContext event) {
+                            @Override
+                            public void handle(RoutingContext event) {
+                                Thread.currentThread().setContextClassLoader(currentCl);
+                                hotReplacementHandler.handle(event);
```
I'm surprised we don't restore the previous CL, here?

marko-bekhta replied:
I can only assume that it's probably so that all subsequent handlers would run with this `currentCl` and it could be  restored in the very last handler, but there's probably not much benefit in that?

cescoffier commented:
+1 for having a dedicated test.

marko-bekhta committed:
Always produce management router

... and decide in the recorder whether we will actually keep that management router or not.

marko-bekhta commented:
I've rebased the PR and marked it as ready-for-review since ... you've been reviewing it 😃
